consensus/event_loop.go

```diff
@@ -469,8 +469,38 @@ func (fc *FastHotStuffEventLoop) onBlockConstructionScheduledTaskExecuted(blockC
 		return
 	}
 
-	// TODO: Check if we can construct a timeout QC for the current view, and send the signal to the
-	// server
+	// Check if we have enough timeouts to build an aggregate QC for the previous view. If so,
+	// we send a signal to the server and cancel all scheduled tasks.
+	if success, previousBlock, highQC, highQCViews, signersList, signature := fc.tryConstructTimeoutQCInCurrentView(); success {
+		// Signal the server that we can construct a timeout QC for the current view
+		fc.ConsensusEvents <- &ConsensusEvent{
+			EventType:   ConsensusEventTypeConstructTimeoutQC, // The event type
+			View:        fc.currentView,                       // The view that we have a timeout QC for
+			BlockHash:   highQC.GetBlockHash(),                // The block hash we can build the aggregate QC from
+			BlockHeight: previousBlock.GetHeight() + 1,        // The block height we can propose a block with this timeout aggregate QC
+			AggregateQC: &aggregateQuorumCertificate{
+				view:        fc.currentView - 1, // The timed out view is always the previous view
+				highQC:      highQC,             // The high QC aggregated from the timeout messages
+				highQCViews: highQCViews,        // The high view for each validator who timed out
+				aggregatedSignature: &aggregatedSignature{
+					signersList: signersList, // The signers list of validators who timed out
+					signature:   signature,   // The aggregated signature from validators who timed out
+				},
+			},
+		}
+
+		// Cancel the block construction task since we know we can construct a timeout QC in the current view.
+		// It will be rescheduled when we advance view.
+		fc.nextBlockConstructionTask.Cancel()
+		return
+	}
+
+	// We have not found a super majority of votes or timeouts. We can schedule the task to check again later.
+	fc.nextBlockConstructionTask.Schedule(
+		fc.blockConstructionInterval,
+		fc.currentView,
+		fc.onBlockConstructionScheduledTaskExecuted,
```

Comment:
Agreed. With the current scheduling, this will re-verify all votes. It is inefficient. But the advantage of the current scheduling is that it's dead simple. And it's likely to be more than sufficient given a crank timer interval of > 1 second, and a validator set size of < 1000 validators.

But the, a more efficient implementation would do this:
- when AdvanceView or ProcessTipBlock is called, it would schedules the crank timer
- on each vote/timeout message, it would keep a cumulative sum of stakes that have voted or timed out per view
- when the crank timer runs, it performs an O(1) lookup of the above stakes for the given view
  - and if there's a supermajority, then it constructs a QC for that view
  - if there isn't a supermajority, then it flips a flag such the next vote/timeout that results in a supermajority triggers the QC construciton
    - then when the next vote/timeout is received that results in a supermajority, it constructs a QC for the view

I'm personally fine running with the dead simple implementation, and keeping the above more efficent process in our back pocket as a later optimization. If we need the optimization, then we can introduce it later as an iteration.